Mobile: select more than one aspect for status message

sebastienadam opened the pull request:
Resolve issue #7080

sebastienadam committed:
#7080 using aspect selecting button form html into mobile version

sebastienadam committed:
Merge branch 'develop' of https://github.com/diaspora/diaspora into develop

sebastienadam committed:
#7080 Adaptation of the JS for the behavior of the aspect selection button.

sebastienadam committed:
7080: SCSS adaptation for displaying the selection of the aspect selection button

sebastienadam committed:
#7080: Cleaning code

sebastienadam committed:
#7080: Updating Cucumber test with new aspect selecting button

svbergerem commented:
Please follow our [styleguide](https://wiki.diasporafoundation.org/Styleguide#Automatic_local_review):

```bash
$ bin/pronto run -c upstream/develop 
app/assets/javascripts/mobile/publisher.js:53 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:56 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:58 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:65 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:78 W: Unexpected space before function parentheses.
app/assets/javascripts/mobile/publisher.js:79 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:79 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:80 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:81 W: Unnecessary semicolon.
app/assets/javascripts/mobile/publisher.js:83 W: Unexpected space before function parentheses.
app/assets/javascripts/mobile/publisher.js:84 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:84 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:85 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:86 W: Unnecessary semicolon.
app/assets/javascripts/mobile/publisher.js:96 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:97 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:98 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:100 W: Infix operators must be spaced.
app/assets/javascripts/mobile/publisher.js:100 W: Infix operators must be spaced.
app/assets/javascripts/mobile/publisher.js:100 W: Infix operators must be spaced.
app/assets/javascripts/mobile/publisher.js:100 W: Infix operators must be spaced.
app/assets/javascripts/mobile/publisher.js:103 W: Unnecessary semicolon.
app/assets/javascripts/mobile/publisher.js:107 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:112 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:115 W: Strings must use doublequote.
app/assets/javascripts/mobile/publisher.js:119 W: There should be no space after '{'
app/assets/javascripts/mobile/publisher.js:119 W: There should be no space before '}'
app/assets/javascripts/mobile/publisher.js:123 W: Strings must use doublequote.
app/assets/stylesheets/mobile/mobile.scss:698 W: Color literals like `#333333` should only be used in variable declarations; they should be referred to via variable everywhere else.
app/assets/stylesheets/mobile/mobile.scss:703 W: Expected item on line 703 to appear before line 697. Rule sets should be ordered as follows: `@extends`, `@includes` without `@content`, properties, `@includes` with `@content`, nested rule sets
app/assets/stylesheets/mobile/mobile.scss:699 W: Rule declaration should be followed by an empty line
app/assets/stylesheets/mobile/mobile.scss:702 W: Rule declaration should be followed by an empty line
app/assets/stylesheets/mobile/mobile.scss:698 W: Color `#333333` should be written as `#333`
app/assets/stylesheets/mobile/mobile.scss:654 W: Avoid using id selectors
app/assets/stylesheets/mobile/mobile.scss:654 W: Avoid using id selectors
app/assets/stylesheets/mobile/mobile.scss:654 W: Avoid using id selectors
app/assets/stylesheets/mobile/mobile.scss:686 W: Nesting should be no greater than 3, but was 4
app/assets/stylesheets/mobile/mobile.scss:687 W: Nesting should be no greater than 3, but was 4
app/assets/stylesheets/mobile/mobile.scss:697 W: Nesting should be no greater than 3, but was 4
app/assets/stylesheets/mobile/mobile.scss:700 W: Nesting should be no greater than 3, but was 4
app/assets/stylesheets/mobile/mobile.scss:691 W: Properties should be ordered display, height, width
app/assets/stylesheets/mobile/mobile.scss:654 W: Avoid qualifying id selectors with an element.
app/assets/stylesheets/mobile/mobile.scss:654 W: Avoid qualifying id selectors with an element.
app/assets/stylesheets/mobile/mobile.scss:697 W: Selector should have depth of applicability no greater than 3, but was 4
app/assets/stylesheets/mobile/mobile.scss:654 W: Selector `user_language` should be written in lowercase with hyphens
app/assets/stylesheets/mobile/mobile.scss:654 W: Selector `user_color_theme` should be written in lowercase with hyphens
app/assets/stylesheets/mobile/mobile.scss:654 W: Selector `user_auto_follow_back_aspect_id` should be written in lowercase with hyphens
app/assets/stylesheets/mobile/mobile.scss:676 W: Selector `aspect_dropdown` should be written in lowercase with hyphens
app/assets/stylesheets/mobile/mobile.scss:690 W: Selector `status_indicator` should be written in lowercase with hyphens
app/assets/stylesheets/mobile/mobile.scss:654 W: Each selector in a comma sequence should be on its own single line
app/assets/stylesheets/mobile/mobile.scss:654 W: Each selector in a comma sequence should be on its own single line
```

Flaburgan commented:
Thank you very much for working on this very asked feature. I tested it locally and it works well. I'm only wondering if it shouldn't be rendered a bit bigger to avoid misclick, as the target is touch screens.

Also a small warning, please do not work in the develop branch next time. You can read [our git workflow](https://wiki.diasporafoundation.org/Git_workflow)

sebastienadam commented:
I will make some adaptation for a better code.

@Flaburgan : I've been a little fuzzy this time, but I'll do better next time.

sebastienadam committed:
#7080: style guide

sebastienadam commented:
I updated the code to be compliant with the style guide. But some adaptations could not be made in existing code.

sebastienadam committed:
#7080: correction in tests

sebastienadam commented:
The tests didn't run with the latest version of the definition. Is it possible to rerun it?

SuperTux88 commented:
They ran, but failed, so you need to fix the tests. Just rerun them would fail again.

sebastienadam commented:
I fixed it, and the good version is in the repository. But the latest run used the previous version.

SuperTux88 commented:
No, the [last run](https://travis-ci.org/diaspora/diaspora/builds/270776407) used [the last commit](https://github.com/diaspora/diaspora/pull/7602/commits/36e165cc4bec1cc88f332acb931fc4babcaccbf6) in this branch. So either you didn't fix or didn't push it.

sebastienadam commented:
OK, so there is a problem I didn't find 🤔 I'll do more investigation.

sebastienadam committed:
Merge branch 'develop' of https://github.com/diaspora/diaspora into develop

sebastienadam committed:
#7080: Correction for selectors

sebastienadam commented:
What a shame! It's not good yet. And this time, the problem is outside the code I modified. 😪 I think I'll retry with a fresh new copy of d* code.

Flaburgan commented:
Sometimes some tests fail randomly. This one looks unrelated to your work. I relaunch the job, to confirm that.

SuperTux88 commented:
Yes, if something breaks related to your code it usually breaks for all ruby versions and all databases (so 4 read jobs or 8 red jobs). If it's only one test that fails once then it's probably a randomly-failing test (especially when you didn't touch code related to that test). And tests are now green after fla restarted the job.

diaspora-code-review commented on app/assets/stylesheets/mobile/mobile.scss:
```diff
@@ -651,7 +651,9 @@ form#update_profile_form {
 }
 
-select#user_language, select#user_color_theme, #user_auto_follow_back_aspect_id, #aspect_ids_ {
+select#user_language,
```
- Avoid using id selectors
- Avoid qualifying id selectors with an element.
- Selector `user_language` should be written in lowercase with hyphens
- Selector `user_color_theme` should be written in lowercase with hyphens
- Selector `user_auto_follow_back_aspect_id` should be written in lowercase with hyphens

sebastienadam commented:
About pronto/scss warning: The problem selectors were already in place. Since they are used outside the code I modified, and I don't know the implications of modifying them, I preferred not to touch them. I just removed a selector that was no longer used in the code I modified.

sebastienadam commented:
Not merged? Something wrong?

Flaburgan commented:
@sebastienadam sorry to take so long to review your work, we're currently [focusing on bug fixes](https://github.com/diaspora/diaspora/commits/release/0.7.1.0) to release the next minor version of diaspora*. We will switch back on new features when diaspora* 0.7.1.0 will be released. Thanks for you patience.

cmrd-senya commented on app/assets/javascripts/mobile/publisher.js:
```diff
@@ -48,4 +48,76 @@ $(document).ready(function(){
 
   new Diaspora.MarkdownEditor("#status_message_text");
+
+  $(".dropdown-menu > li").bind("tap click", function(evt) {
```
The code you add here is mostly copy-pasted code from [`app/assets/javascripts/app/views/publisher/aspect_selector_view.js`](https://github.com/diaspora/diaspora/blob/b263cf7e6db3653e6751a47526c941d5c4518910/app/assets/javascripts/app/views/publisher/aspect_selector_view.js#L21). Such approach in programming is highly discouraged. Instead you should create a helper class which could be shared between mobile and desktop versions, like we have it for the [markdown editor](https://github.com/diaspora/diaspora/blob/develop/app/assets/javascripts/helpers/markdown_editor.js).

Flaburgan replied:
@sebastienadam can you please have a look at that remark? Then we will do the review ;)

cmrd-senya replied:
I could try to fix this myself too, but I better trust the judgement of Denis here

tclaus commented:
To get  new status on this: 
This feature is IMHO still very requested - just today I wondered how to select multiple aspects in mobile view. Then I stumbled to this PR. After a rebase to current development I found out this this works well for mobile views. 
It changes browser behavior from OS-Specific to the bootstrap one, which I like more. 
Here my current test-results: 

Mobile View on device
<img width="434" alt="Bildschirmfoto 2021-02-28 um 17 49 34" src="https://user-images.githubusercontent.com/501326/109427419-a27bc900-79f2-11eb-9599-15d7e0d087c3.png">

Mobile view on Desktop
<img width="496" alt="Bildschirmfoto 2021-02-28 um 17 49 20" src="https://user-images.githubusercontent.com/501326/109427425-a871aa00-79f2-11eb-9912-08d444b7dcd9.png">

The existing desktop views still run. 
I can adopt this PR, rebase it and create a new PR. There is also the "Scrollbars on Aspects" PR #8213

Flaburgan commented:
Superseded by https://github.com/diaspora/diaspora/pull/8217, closing this one.
